### ISSUE TITLE
JS: Enforce trailing semicolons?

### ISSUE THREAD

**Issue by sambulance** [humanmade/coding-standards#153]
Our linter doesn't pick up on trailing semicolons. This can lead to inconsistency inside JS files:

```javascript
const showSection = section => {
	section.setAttribute( 'aria-expanded', true )
}

const hideSection = section => {
	section.setAttribute( 'aria-expanded', false );
};
```

Both of the above are valid. We should enforce either one way or the other.

**Comment by mikeselander** [comment]
This is something that we should get general consensus on since it's more hot-button. @humanmade/frontend any thoughts here?

**Comment by ntwb** [comment]
Not explicitly mentioned in the handbook, but all JS examples use semi-colons:

https://engineering.hmn.md/standards/style/js/

**Comment by mattheu** [comment]
Are there any reasons not to enforce this?

ESLint should be pretty good at automatically fixing issues in old code.

**Comment by fklein-lu** [comment]
I consider that enforcing trailing semicolons is the best approach.

First of all, much like in PHP, `;` ends a statement. Which is unlike with automatic semicolon insertion. Here the rule is `\n` ends a statement, unless this and that and the other.

Second but there are common situations in JavaScript that work properly without semicolons. For example `( function () { [...] }() )` would need a prepending `;`.

There can also be issues when code is minified, as the automatic semicolon insertion relies on white space. 

For reference:
- https://google.github.io/styleguide/jsguide.html#formatting-semicolons-are-required
- https://github.com/airbnb/javascript#semicolons

**Comment by mattheu** [comment]
I've opened a PR to add this rule. See #169

**Comment by goldenapples** [comment]
Unrelated side note: I was actually surprised recently to learn that `phpcs` doesn't enforce trailing semicolons where it's not required by the language: both `<?php esc_html_e( 'foo' ) ?>` or `<?= esc_html( $var ) ?>` pass with HM Coding Standards.

**Comment by sambulance** [comment]
@goldenapples Our guidelines explicitly say to drop the semicolon for inline statements in PHP: https://engineering.hmn.md/standards/style/php/#inline-statements 

I'm also in favour of enforcing, thanks for gathering feedback and pushing ahead @mikeselander & @mattheu

**Comment by rmccue** [comment]
> First of all, much like in PHP, `;` ends a statement. Which is unlike with automatic semicolon insertion. Here the rule is `\n` ends a statement, unless this and that and the other.

The counter-argument to this is that automatic semicolon insertion happens regardless of whether you put semicolons in or not, so you have to learn the rules anyway.

I think we should enforce this for consistency with PHP.

(And yes, inline statements in PHP don't require semicolons; that's also consistent with something like `<p>{ text }</p>` in React, although we don't get a choice there.)

**Comment by goldenapples** [comment]
> Our guidelines explicitly say to drop the semicolon for inline statements in PHP

Whoa, I've missed that guideline all this time!
